Background check of certificate states (EXPOSUREAPP-8364)

d4rken opened the pull request:
Wires up most  of the state check background logic.

~~- [ ] Implement actually showing notifications (may be done in a follow up PR)~~
- [x] Fill out unit tests (may be done in a follow up PR)

chiljamgossow committed:
check state

chiljamgossow committed:
Merge branch 'release/2.7.x' into feature/8369-exp-check-worker

chiljamgossow committed:
move to new package

chiljamgossow committed:
Merge branch 'release/2.7.x' into feature/8369-exp-check-worker

chiljamgossow committed:
move to new package

d4rken committed:
Trigger fore/background notifications based on certificate expiration status.

d4rken committed:
Merge remote-tracking branch 'origin/feature/8369-exp-check-worker' into feature/8364-dsc-refresh-scheduler

# Conflicts:
#	Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/common/statecheck/DccStateCheckWork.kt
#	Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/expiration/DccExpirationNotification.kt
#	Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/signature/core/execution/DccStateCheckWorker.kt

harambasicluka committed:
Merge branch 'release/2.7.x' into feature/8364-dsc-refresh-scheduler

d4rken committed:
Include #3787

d4rken committed:
Lints

d4rken committed:
Merge remote-tracking branch 'origin/feature/8364-dsc-refresh-scheduler' into feature/8364-dsc-refresh-scheduler

d4rken committed:
Cleanup

d4rken committed:
More wiring signature and expiration check.

d4rken commented on Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/signature/core/DscData.kt:
```diff
@@ -1,7 +1,9 @@
 data class DscData(
-    val dscList: List<Pair<ByteString, ByteString>>
+    val dscList: List<Pair<ByteString, ByteString>>,
+    val updatedAt: Instant,
```
@jurajkusnier You are doing the repo right? Could you provide such an `updatedAt` field?

jurajkusnier replied:
sure, I can add it
btw I will change `dscList` type to `List<DscListItem>` instead of `Pair<>`

jurajkusnier replied:
what should be in `updatedAt` when we load data from `DefaultDscSource` ?

d4rken replied:
`Instant.EPOCH`

chiljamgossow commented on Corona-Warn-App/src/main/java/de/rki/coronawarnapp/appconfig/mapping/CovidCertificateConfigMapper.kt:
```diff
@@ -18,7 +18,9 @@ class CovidCertificateConfigMapper @Inject constructor() : CovidCertificateConfi
         return CovidCertificateConfigContainer(
             testCertificate = rawConfig.dgcParameters.mapCovidCertificateConfig(),
-            expirationThresholdInDays = rawConfig.dgcParameters.expirationThresholdInDays
+            expirationThresholdInDays = Duration.standardDays(
```
Max said this is the number of date changes, not a multiple of 24h intervals. That's why I think this conversion is not accurate, however, it won't have any meaningful consequences for users.

d4rken replied:
Seems like a duration to me though, whether we treat it as day changes would be up to the configs consumer. I'd prefer to model `expirationThresholdInDays` as a Duration.
I'll rename it `expirationThreshold` as whether it is hours/days/weeks is encoded in `Duration` and not something the config consumer has to care about.

chiljamgossow commented on Corona-Warn-App/src/main/java/de/rki/coronawarnapp/appconfig/mapping/CovidCertificateConfigMapper.kt:
```diff
@@ -51,7 +53,7 @@ class CovidCertificateConfigMapper @Inject constructor() : CovidCertificateConfi
 
     data class CovidCertificateConfigContainer(
         override val testCertificate: CovidCertificateConfig.TestCertificate = TestCertificateConfigContainer(),
-        override val expirationThresholdInDays: Int = 1
```
I put that in here deliberately, not sure what the fallback should be.

d4rken replied:
I changed it to 14 days based on:
https://github.com/corona-warn-app/cwa-app-android/pull/3799#discussion_r676476499
and added some unit tests.

chiljamgossow commented on Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/common/statecheck/DccStateChecker.kt:
```diff
@@ -0,0 +1,46 @@
+        val nowUtc = timeStamper.nowUTC
+
+        // expiration check
+        val expirationState = expirationChecker.getExpirationState(
```
you could move that down and only check expiration for valid signatures

d4rken replied:
Thanks, good point!

mtwalli commented on Corona-Warn-App/src/main/java/de/rki/coronawarnapp/appconfig/mapping/CovidCertificateConfigMapper.kt:
```diff
@@ -51,7 +53,7 @@ class CovidCertificateConfigMapper @Inject constructor() : CovidCertificateConfi
     data class CovidCertificateConfigContainer(
         override val testCertificate: CovidCertificateConfig.TestCertificate = TestCertificateConfigContainer(),
-        override val expirationThresholdInDays: Int = 1
+        override val expirationThresholdInDays: Duration = Duration.standardDays(1),
```
should not be `14` ? https://github.com/corona-warn-app/cwa-app-tech-spec/blob/78009f7e3aa6aef56892caedcfe478bfb9d4d691/docs/spec/handling-of-app-config.md#parameters-for-digital-green-certificate

d4rken replied:
Done and added tests.

mtwalli commented on Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/expiration/DccExpirationChecker.kt:
```diff
@@ -0,0 +1,31 @@
+        val daysUntilExpiration = now.daysUntil(expiresAt, timeZone)
+        val diff = daysUntilExpiration - expirationThreshold.standardDays
+        return when {
+            daysUntilExpiration <= 0 -> CwaCovidCertificate.State.Expired(expiresAt)
```
Something to consider in unit tests?
https://github.com/corona-warn-app/cwa-app-android/pull/3804

d4rken committed:
Unit tests.

d4rken committed:
Refactoring based on PR comments.

mtwalli commented on Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/common/statecheck/DccStateChecker.kt:
```diff
@@ -0,0 +1,49 @@
+        if (!isSignatureValid) {
+            Timber.tag(TAG).w("Certificate had invalid signature: %s (kid)", dccData)
+            return@combine CwaCovidCertificate.State.Invalid
+        }
```
`DscSignatureValidator` does not return boolean anymore , you should wrap`dscSignatureValidator.isSignatureValid(dscData, dccData)`  in a` try {} catch {}` 
if it throw an `InvalidHealthCertificateException` -> `Invalid` ,otherwise it should be `Valid`

d4rken replied:
Done
https://github.com/corona-warn-app/cwa-app-android/pull/3799/commits/256de94211174800a2c7a04ac3218e26a1e55535

d4rken committed:
Adjust expiration state calculation based on #3804

d4rken committed:
Refactoring

d4rken committed:
Adjust signature validation step to expect failure via exception.
https://github.com/corona-warn-app/cwa-app-android/pull/3799#discussion_r676674348

d4rken committed:
Fix day check not working if just the year changes.

d4rken committed:
Merge branch 'release/2.7.x' into feature/8364-dsc-refresh-scheduler

chiljamgossow commented on Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/expiration/DccExpirationChecker.kt:
```diff
@@ -0,0 +1,38 @@
+        return when {
+            daysUntilExpiration < 0 -> CwaCovidCertificate.State.Expired(expiresAt)
+            daysUntilExpiration == 0 -> {
+                if (now.toDateTime(timeZone).isAfter(expiresAt)) {
```
Here you can just compare millis, both are Instants. time zone should not affect this.

d4rken replied:
Thanks, will adjust.

chiljamgossow commented on Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/signature/core/DscSignatureValidator.kt:
```diff
@@ -8,10 +8,9 @@ import javax.inject.Inject
 class DscSignatureValidator @Inject constructor() {
 
-    suspend fun isSignatureValid(dscData: DscData, certificateData: DccData<*>): Boolean {
+    suspend fun isSignatureValid(dscData: DscData, certificateData: DccData<*>) {
```
As it does not return a Boolean any more, is... is misleading. Maybe just checkSignature(...)?

d4rken replied:
Was changed to `validateSignature` by merging 2.7 into my branch.

chiljamgossow commented on Corona-Warn-App/src/main/java/de/rki/coronawarnapp/util/TimeAndDateExtensions.kt:
```diff
@@ -112,6 +112,10 @@ object TimeAndDateExtensions {
     fun Instant.toLocalDateUserTz(): LocalDate = this.toUserTimeZone().toLocalDate()
 
+    fun LocalDate.isDifferentDay(other: LocalDate): Boolean {
+        return this.dayOfYear != other.dayOfYear || this.year != other.year
```
Equals does not work?

d4rken replied:
I thought this wouldn't work, but you are right, it does work. Removed the extension function again.

harambasicluka committed:
Merge branch 'release/2.7.x' into feature/8364-dsc-refresh-scheduler

d4rken committed:
Merge branch 'release/2.7.x' into feature/8364-dsc-refresh-scheduler

# Conflicts:
#	Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/signature/core/DscData.kt
#	Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/signature/core/DscSignatureValidator.kt
#	Corona-Warn-App/src/test/java/de/rki/coronawarnapp/covidcertificate/CovidCertificateTestComponent.kt

d4rken committed:
Integrate PR feedback.

chiljamgossow reviewed (approved):
Thanks!

mtwalli committed:
Merge branch 'release/2.7.x' into feature/8364-dsc-refresh-scheduler

mtwalli committed:
Merge branch 'release/2.7.x' into feature/8364-dsc-refresh-scheduler
